### PR TITLE
implemented report and warnings system

### PR DESCRIPTION
**Pull request by joewadk**
- made changes to tables.Post to allow "REPORTS" and "DISPUTE" type posts
- made changes to tables.User to track fines, time of last warn, and number of warnings, 
- added in posts route 5 different endpoints for reportpage, reports, report dispute, report approve, and report disapprove
- added in users route warn endpoint, allowing super user to warn other users
- modified authenticate function in common to make sure fines!=0, warnings<3, and time between current time and time of last warn<5 days so that the user can continue using the website.

### CODE REVIEW

**Comment by DUOLabs333** [comment]
Is the node_modules folder needed? And resolve the conflicts (I made some updates and I want to make sure all code is up-to-date).

**Comment by joewadk** [comment]
ok conflicts should be good now

**Comment by DUOLabs333** [comment]
Some Git merging artifacts are still there (eg, `<<<<<<< HEAD`), and the `node_modules` folder is still there.

**Comment by joewadk** [comment]
artifacts removed and node_modules folder removed

File: routes/posts.py [hunk 1]
```diff
@@ -11,8 +11,11 @@
 import base64, os, random, string
 import multiprocessing
 from pathlib import Path
+import datetime from datetime
+lock=multiprocessing.Lock() #Not enough to use autoincrement ---- autoincrement doesn't neccessarily create monotonically increasing IDs, only unique ones. However, we need it in a specific order.
```
Comment (DUOLabs333) [inline routes/posts.py]: Not needed.

File: routes/posts.py [hunk 2]
```diff
@@ -11,8 +11,11 @@
...
 
-@app.route("/posts/homepage")
+
+@app.route("/posts/homepage", methods = ['POST'])
```
Comment (DUOLabs333) [inline routes/posts.py]: /homepage has been rewritten. Make sure your branch has incorporated all changes from main so unnecessary changes won't show up in the diff.

File: routes/posts.py [hunk 3]
```diff
@@ -23,15 +26,37 @@ def homepage():
...
-        result["before"]=before+len(result["posts"])
+        post_query=select(tables.Post.id).where(user.has_followed(tables.Post.author) & (tables.Post.id < before) & not_(user.has_blocked(tables.Post.author)) & (tables.Post.type=="POST") ).limit(limit).order_by(desc(tables.Post.id)) #Sort chronologically, not algorithmically --- one of the biggest problems with other social media sites
+       
+        posts = session.scalars(post_query).all()  # Fetch regular posts
+        result["posts"]=session.scalars(posts).all()
         
         return result
+    
+
+@app.route("/posts/reportpage", methods = ['POST'])
```
Comment (DUOLabs333) [inline routes/posts.py]: Move all report-functionality to its own file --- 'routes/report.py`

File: routes/posts.py [hunk 4]
```diff
@@ -337,6 +362,137 @@ def image():
         return send_file(path, mimetype=type)
         
     
+@app.route("/posts/reports", methods=['POST'])
```
Comment (DUOLabs333) [inline routes/posts.py]: Should just be /posts/report (unlike what I said above, keep this endpoint in this file).

File: routes/posts.py [hunk 5]
```diff
@@ -337,6 +362,137 @@ def image():
...
+        error, data = posts.cleanPostData(None, report_data, complainer)  #filter taboo
+        if error!=None:
+            result["error"]=error
+            return
+            
+        # Create the report post
+        result["id"] = posts.createPost(data)
+            
+
+@app.route("/posts/reports/dispute", methods=['POST'])
```
Comment (DUOLabs333) [inline routes/posts.py]: Move this endpoint to the reports.py file mentioned above.

File: routes/posts.py [hunk 6]
```diff
@@ -337,6 +362,137 @@ def image():
...
+@common.authenticate        #this endpoint feels a little weird, i might have to work on it more
+def dispute_report():
+    result = {}
     
+    uid = request.json["uid"] #the user id for complainee
+    report_id = request.json["report_id"]  # ID of the report being disputed
+    dispute_text = request.json["dispute_text"]  # The text the complainee adds to dispute the report
```
Comment (DUOLabs333) [inline routes/posts.py]: Rename to "dispute_reason"

File: routes/posts.py [hunk 7]
```diff
@@ -337,6 +362,137 @@ def image():
...
     
-    
\ No newline at end of file
+    with Session(common.database) as session:
+        complainee = users.getUser(uid, session) 
+        report = session.query(tables.Post).filter(tables.Post.id == report_id)
```
Comment (DUOLabs333) [inline routes/posts.py]: Don't use query --- use select, scalars, then use `.first()`.

File: routes/posts.py [hunk 8]
```diff
@@ -337,6 +362,137 @@ def image():
...
+
+        # Ensure the report exists and the complainee is the one being reported
+        if  report.author == complainee.id and report.type == "DISPUTE":
+            # Update the report text with the dispute information
+            report.text += "\n\nDispute by User (ID: {}):\n{}".format(complainee.id, dispute_text)
+            session.commit()
+            result["message"] = "Report dispute has been recorded."
```
Comment (DUOLabs333) [inline routes/posts.py]: Don't return a message if successful.

File: routes/posts.py [hunk 9]
```diff
@@ -337,6 +362,137 @@ def image():
...
+        else:
+            result["error"] = "Report not found or permission denied."
```
Comment (DUOLabs333) [inline routes/posts.py]: All errors must be capitalized, with no spaces.

File: routes/posts.py [hunk 10]
```diff
@@ -337,6 +362,137 @@ def image():
...
+    report_id = request.json["report_id"]  # ID of the report being disputed
+    dispute_text = request.json["dispute_text"]  # The text the complainee adds to dispute the report
     
-    
\ No newline at end of file
+    with Session(common.database) as session:
+        complainee = users.getUser(uid, session) 
+        report = session.query(tables.Post).filter(tables.Post.id == report_id)
+
+        # Ensure the report exists and the complainee is the one being reported
```
Comment (DUOLabs333) [inline routes/posts.py]: The report.type is "REPORT", not "DISPUTE". This endpoint should end up adding a new DISPUTE post, not modifying an existing dispute.

File: routes/posts.py [hunk 11]
```diff
@@ -337,6 +362,137 @@ def image():
...
+            session.commit()
+            result["message"] = "Report dispute has been recorded."
+        else:
+            result["error"] = "Report not found or permission denied."
+            
+    
+    return result
+
+
+@app.route("/posts/reports/approve", methods=['POST'])
```
Comment (DUOLabs333) [inline routes/posts.py]: Move endpoint to new file.

File: routes/posts.py [hunk 12]
```diff
@@ -337,6 +362,137 @@ def image():
...
+        if not super_user.hasType(User.SUPER):
+            result["error"] = "INSUFFICIENT_PERMISSION"
+            return result
+        
+        report = session.query(tables.Post).filter(tables.Post.id == report_id)
+
+        # Verify the report exists and it's of type 'REPORT'
+        if report and report.type == "REPORT":
+            complainee = users.getUser(target_user, session)  # Get the user being reported
+            complainee.warnings += 1  # Increment the warnings count
```
Comment (DUOLabs333) [inline routes/posts.py]: Do not add a new field, just post a comment of type WARNING to the user's inbox.

File: routes/posts.py [hunk 13]
```diff
@@ -337,6 +362,137 @@ def image():
...
+            complainee.time_of_last_warn = datetime.utcnow()  # add warning based on current time
```
Comment (DUOLabs333) [inline routes/posts.py]: See above.

File: routes/posts.py [hunk 14]
```diff
@@ -337,6 +362,137 @@ def image():
...
+
+            session.commit()
+            result["message"] = "Report has been approved and warning issued."
```
Comment (DUOLabs333) [inline routes/posts.py]: No message is neeeded.

File: routes/posts.py [hunk 15]
```diff
@@ -337,6 +362,137 @@ def image():
...
+        else:
+            result["error"] = "Report not found." #can't find report
+
```
Comment (DUOLabs333) [inline routes/posts.py]: Remember to delete report.

File: routes/posts.py [hunk 16]
```diff
@@ -337,6 +362,137 @@ def image():
...
+    return result
+
+@app.route("/posts/reports/disapprove", methods=['POST'])
```
Comment (DUOLabs333) [inline routes/posts.py]: Similar to above.

File: routes/posts.py [hunk 17]
```diff
@@ -337,6 +362,137 @@ def image():
...
+        if not super_user.hasType(User.SUPER):
+            result["error"] = "INSUFFICIENT_PERMISSION"
+            return result
+        
+        report = session.query(tables.Post).filter(tables.Post.id == report_id)
+        # Verify the report exists and it's of type 'REPORT'
+        if report and report.type == "REPORT":
+
+
+@app.route("/posts/reports/disapprove", methods=['POST'])    #mostly a copy of /posts/delete
```
Comment (DUOLabs333) [inline routes/posts.py]: I feel like this is an artifact from a mistake when merging.

File: routes/search.py [hunk 18]
```diff
@@ -57,5 +57,9 @@ def search():
         query=select(tables.Post.id).where(authors & or_(*keywords) & (tables.Post.likes >= likes[0]) & (tables.Post.likes <= likes[1]) & (tables.Post.dislikes >= dislikes[0]) & (tables.Post.dislikes <= dislikes[1]) & not_(func.has_blocked(tables.Post.author)) & tables.Post.is_viewable(user) & tables.Post.type.in_(types)).order_by(sort.desc()).offset(before).limit(limit) #Order by number of keywords satisfied
         
         result["posts"]=session.scalars(query).all()
+<<<<<<< HEAD
```
Comment (DUOLabs333) [inline routes/search.py]: Again, make sure you merged main.

File: routes/users.py [hunk 19]
```diff
@@ -393,3 +393,31 @@ def suggest():
     result["users"]=[_[0] for _ in heapq.nlargest(min(10,len(list_of_users)),list_of_users,key=lambda x: x[1])]
     
     return result
+
+
+
+@app.route("/users/warn", methods=["POST"])
```
Comment (DUOLabs333) [inline routes/users.py:399]: Approving a report automatically sends a warning. There is no need for this endpoint.

File: utils/common.py [hunk 20]
```diff
@@ -53,9 +53,9 @@ def wrapper(*args,**kwargs):
        
        user=users.getUser(uid)
        
-       has_access=(user is not None) and (hash==user.password_hash)
-       
-       if not has_access:
+       has_access=(user is not None) and (hash==user.password_hash)
```
Comment (DUOLabs333) [inline utils/common.py]: Query to get number of warnings and time of last warning --- don't use bespoke fields.

File: utils/common.py [hunk 21]
```diff
@@ -53,9 +53,9 @@ def wrapper(*args,**kwargs):
        
        user=users.getUser(uid)
        
-       has_access=(user is not None) and (hash==user.password_hash)
-       
-       if not has_access:
+       has_access=(user is not None) and (hash==user.password_hash) 
+       is_deleted=(user.warnings>3) and ((datetime.utcnow-user.time_of_last_warn).days)>5 and (user.fines!=0)
+       if (not has_access or is_deleted ) :                #user does not have access or is deleted
```
Comment (DUOLabs333) [inline utils/common.py]: You forgot to actually delete the user.

File: utils/tables.py [hunk 22]
```diff
@@ -28,7 +29,10 @@ class User(BaseTable):
     inbox: Mapped[int] = mapped_column(default=0)
     profile: Mapped[int] = mapped_column(default=0)
     applied: Mapped[int] = mapped_column(default="")
-    
+    fines: Mapped[int] = mapped_column(default=0)  #added fines for user
+    warnings: Mapped[int] = mapped_column(default=0)
```
Comment (DUOLabs333) [inline utils/tables.py:33]: These attributes are not needed.

File: utils/tables.py [hunk 23]
```diff
@@ -133,7 +137,8 @@ def is_trendy(self):
     @hybrid_property
     def trendy_ranking(self):
         return self.views/(self.dislikes+1)
-    
+        
+    public_post_types=["JOB","AD","POST","COMMENT", "REPORT", "DISPUTE"]
```
Comment (DUOLabs333) [inline utils/tables.py]: REPORT and DISPUTE should not be public.

File: utils/users.py [hunk 24]
```diff
@@ -49,4 +50,60 @@ def update_trendy_status(user):
             user.addType(user.TRENDY)
         else:
             user.removeType(user.TRENDY)
-        session.commit()
\ No newline at end of file
+        session.commit()
+
+def get_user_warnings_count(user):    #returns warnings for any given user
```
Comment (DUOLabs333) [inline utils/users.py:55]: Probably a mistake when merging.

File: utils/users.py [hunk 25]
```diff
@@ -49,4 +50,60 @@ def update_trendy_status(user):
...
+    with Session(common.database) as session:
+            warnings_count= user.warnings
+            return warnings_count
+    
+def handle_three_warnings(user):
```
Comment (DUOLabs333) [inline utils/users.py]: Make a separate endpoint for handling user paying to get rid of the warnings instead of whatever this is.

**Review by DUOLabs333 — COMMENTED** [review]
You have your hands full.

Also, delete the .vscode, package.json and package-lock.json

**Comment by joewadk** [comment]
ok reports.py is live, in addition most error comments have been resolved

**Comment by DUOLabs333** [comment]
This PR has been successfully incorporated into the main branch.
